route_guide/route_guide_client.cc

```diff
@@ -23,238 +23,346 @@
+    // in order to have the way of injecting some data on a stack, or as
+    // a part of the eventual data structure and manipulate with this data in
+    // multiple callbacks (.start(), .fail(), .ready(), .body(), etc).
+    struct Data {
+      Point point; // Request from the client.
+      Feature feature; // Response from the server.
+    };
 
-    Notification<Status> done;
-    client_->Call<RouteGuide, Rectangle, Stream<Feature>>(
-        "ListFeatures",
-        &rect,
-        [this](auto* call, auto&& feature) {
-          if (feature) {
-            std::cout << "Found feature called "
-                      << feature->name() << " at "
-                      << feature->location().latitude()/kCoordFactor_ << ", "
-                      << feature->location().longitude()/kCoordFactor_ << std::endl;
-          } else {
-            call->Finish();
-          }
-        },
-        [&](auto*, const Status& status) {
-          done.Notify(status);
-        });
+    // Call() is an `Eventual` (a concept from `stout-eventuals`) and can
+    // be composed into an asynchronous pipeline with other Eventuals.
+    // In this example we use the "|" operator to compose it with a
+    // `Client::Handler()`. This function gets the remote function's name
+    // as a parameter. The handler in turn could be composed also.
+    // For more information about Eventuals, see:
+    // https://github.com/3rdparty/stout-eventuals.
+    return client_->Call<RouteGuide, Point, Feature>("GetFeature")
+        | (Client::Handler()
+               // Passing Data structure to the context.
+               .context(Data{MakePoint(latitude, longitude), Feature{}})
+               // Send the request with Point's data to the server. First
+               // argument in the .ready() callback gets the context. The
+               // second argument gets the `call` object, which is how we
+               // interact with the running RPC. You can send requests to
+               // the server using Write*() family functions. In this
+               // case we just send a unique request and finish our call
+               // with WriteLast function.
+               .ready([](auto&& data, auto& call) {
+                 call.WriteLast(data.point);
+               })
+               // The `.body()` callback is invoked every time the server
+               // sends a response. The lambda's first argument gets the
+               // context. The second gets the call. The third gets the
+               // response object from the server.
+               .body([this](auto&& data, auto& call, auto&& response) {
+                 // Do not forget to check if the response from the server
+                 // is valid! If it is, we process the response. Remember that
+                 // all streams end with a final invalid response, even those
+                 // that finish correctly!
+                 if (!response) {
```

Comment:
As with the comments below about type-safe checks:

```suggestion
                 if (response != nullptr) {
```